test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,352 @@
+    EXPECT_NE(handle, nullptr);
+    server_signal->WaitForNotification();
+    EXPECT_NE(server_endpoint, nullptr);
+    ++g_num_active_connections;
+    connections.push_back(
+        std::make_tuple<std::unique_ptr<Endpoint>, std::unique_ptr<Endpoint>>(
+            CreatePosixEndpoint(handle,
+                                PosixEngineClosure::TestOnlyToClosure(
+                                    [poller](absl::Status /*status*/) {
+                                      if (--g_num_active_connections == 0) {
+                                        poller->Kick();
+                                      }
+                                    }),
+                                posix_ee, config),
+            std::move(server_endpoint)));
+    delete server_signal;
+    server_signal = new grpc_core::Notification();
+  }
+  delete server_signal;
+  return connections;
+}
+
+}  // namespace
+
+class TestParam {
+ public:
+  TestParam(std::string poller, bool is_zero_copy_enabled)
+      : poller_(std::move(poller)),
+        is_zero_copy_enabled_(is_zero_copy_enabled) {}
+
+  std::string Poller() const { return poller_; }
+  bool IsZeroCopyEnabled() const { return is_zero_copy_enabled_; }
+
+ private:
+  std::string poller_;
+  bool is_zero_copy_enabled_;
+};
+
+std::string TestScenarioName(const ::testing::TestParamInfo<TestParam>& info) {
+  return absl::StrCat("poller_type_", info.param.Poller(),
+                      "_is_zero_copy_enabled_", info.param.IsZeroCopyEnabled());
+}
+
+// A helper class to drive the polling of Fds. It repeatedly calls the Work(..)
+// method on the poller to get pet pending events, then schedules another
+// parallel Work(..) instantiation and processes these pending events. This
+// continues until all Fds have orphaned themselves.
+class Worker : public grpc_core::DualRefCounted<Worker> {
+ public:
+  Worker(std::shared_ptr<EventEngine> engine, PosixEventPoller* poller)
+      : engine_(std::move(engine)), poller_(poller) {
+    WeakRef().release();
+  }
+  void Orphan() override { signal.Notify(); }
+  void Start() {
+    // Start executing Work(..).
+    engine_->Run([this]() { Work(); });
+  }
+
+  void Wait() {
+    signal.WaitForNotification();
+    WeakUnref();
+  }
+
+ private:
+  void Work() {
+    auto result = poller_->Work(24h, [this]() {
+      // Schedule next work instantiation immediately and take a Ref for
+      // the next instantiation.
+      Ref().release();
+      engine_->Run([this]() { Work(); });
+    });
+    ASSERT_TRUE(result == Poller::WorkResult::kOk ||
+                result == Poller::WorkResult::kKicked);
+    // Corresponds to the Ref taken for the current instantiation. If the
+    // result was Poller::WorkResult::kKicked, then the next work instantiation
+    // would not have been scheduled and the poll_again callback would have
+    // been deleted.
+    Unref();
+  }
+  std::shared_ptr<EventEngine> engine_;
+  PosixEventPoller* poller_;
```

Comment:
poller is unowned here, right?

Reply:
Yes its not owner by worker. Rather its owned by the test which creates an instance of Worker

Reply:
It's probably worth documenting the ownership and lifetime expectations since the type system isn't going to enforce it.

Reply:
Added comment now.